src/System.Slices/src/System/SpanEnumerators.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System
+{
+    public partial struct Span<T>
+    {
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// </summary>
+        public struct Enumerator
+        {
```

Comment:
The other subtle thing here is that IEnumerator has stronger requirements for MoveNext. Implementation must continue returning "false" from MoveNext after iteration is over, no matter how many times called. (think of the overflowing the index). 
Foreach pattern does not impose such requirements. Since compiler will not call MoveNext redundantly, it is possible to have simpler and more efficient implementation.

If performance is super important, and it is in this case since we are dealing with fundamental scenario competing with array indexing, it would be reasonable to have separate enumerators.